flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java

```diff
@@ -0,0 +1,184 @@
+        this.rowType = checkNotNull(rowType);
+        this.serializedData = checkNotNull(serializedData);
+        this.isFiltering = isFiltering;
+    }
+
+    public boolean isFiltering() {
+        return isFiltering;
+    }
+
+    public RowType getRowType() {
+        return rowType;
+    }
+
+    public boolean contains(RowData row) {
+        if (!isFiltering) {
+            return true;
+        } else if (row.getArity() != rowType.getFieldCount()) {
+            throw new TableException("The arity of RowData is different");
+        } else {
+            prepare();
+            List<RowData> mayMatchRowData = dataMap.get(hash(row));
+            if (mayMatchRowData == null) {
+                return false;
+            }
+            for (RowData mayMatch : mayMatchRowData) {
+                if (matchRow(row, mayMatch)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    private boolean matchRow(RowData row, RowData mayMatch) {
+        for (int i = 0; i < rowType.getFieldCount(); ++i) {
+            if (!Objects.equals(
+                    fieldGetters[i].getFieldOrNull(row),
+                    fieldGetters[i].getFieldOrNull(mayMatch))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void prepare() {
+        if (!prepared) {
+            synchronized (this) {
+                if (!prepared) {
+                    doPrepare();
+                    prepared = true;
+                }
+            }
+        }
+    }
+
+    private void doPrepare() {
+        this.dataMap = new HashMap<>();
+        if (isFiltering) {
+            this.fieldGetters =
+                    IntStream.range(0, rowType.getFieldCount())
+                            .mapToObj(i -> RowData.createFieldGetter(rowType.getTypeAt(i), i))
+                            .toArray(RowData.FieldGetter[]::new);
+
+            TypeSerializer<RowData> serializer = typeInfo.createSerializer(new ExecutionConfig());
+            for (byte[] bytes : serializedData) {
+                try (ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
+                        DataInputViewStreamWrapper inView = new DataInputViewStreamWrapper(bais)) {
+                    RowData partition = serializer.deserialize(inView);
+                    List<RowData> partitions =
+                            dataMap.computeIfAbsent(hash(partition), k -> new ArrayList<>());
+                    partitions.add(partition);
+                } catch (Exception e) {
+                    throw new TableException("Unable to deserialize the value.", e);
+                }
+            }
+        }
+    }
+
+    private int hash(RowData row) {
+        return Objects.hash(Arrays.stream(fieldGetters).map(g -> g.getFieldOrNull(row)).toArray());
+    }
+
+    public static boolean isEqual(DynamicFilteringData data, DynamicFilteringData another) {
```

Comment:
implements `equals` method directly ?

Reply:
As the data struct may contain large scale of data, calculate the hashcode with all the data can be costing. So I suppose we'd better not to override the hashcode, as well as the equals.